### ISSUE TITLE
How would I check symlinks in ext4img.

### ISSUE THREAD

**Issue by pr0f35510n4l** [cubinator/ext4#11]
Hello, For the past week i've been trying to read all symlinks and where they point to using this tool.

But I never succeeded. How can I do that?

**Comment by cubinator** [comment]
Symbolic links are just normal files whose content is the target path. If you've got a any `ext4.Inode` object (that represents a symbolic link), e.g. `file = volume.root.get_inode("path/to/some/symlink")` or `file = volume.get_inode(inode_idx)`, you can simply read it like a normal file `file.open_read().read().decode("utf8")`.

To obtain all the symlinks, you could iterate the root directory recursively:

    import ext4

    def find_symlinks (volume):
        def scan_dir (directory):
            for file_name, inode_idx, file_type in directory.open_dir():
                if file_name in (".", ".."): continue

                inode = volume.get_inode(inode_idx)

                if file_type == ext4.InodeType.DIRECTORY:
                    yield from ((f"{file_name}/{file_name2}", target) for file_name2, target in scan_dir(inode))
                elif file_type == ext4.InodeType.SYMBOLIC_LINK:
                    target = inode.open_read().read().decode("utf8")
                    yield (file_name, target)

        yield from scan_dir(volume.root)

    with open("path/to/some/ext4_image", "rb") as file:
        volume = ext4.Volume(file)

        for path, target in find_symlinks(volume):
            print(f"{path} -> {target}")

**Comment by cubinator** [comment]
While testing the `find_symbolics` function above I noticed a problem\* with how my module checks whether a given inode is a directory or a file. I uploaded a new version of the module, where this is should be fixed now.

---

\* Essentially, the ext4 format stores the file type of any given inode in two places, in the inode itself (`ext4.ext4_inode.i_mode`) and in the directory entry (`ext4.ext4_dir_entry_2.file_type`) of the directory containing the inode. For whatever reason I implemented the file type checks `ext4.Inode.is_file` and `ext4.Inode.is_dir` in such a way that they gave precedence to the directory entry's file type instead of the inode's own file type. The problem arises when the inode was obtained via `ext4.Volume.get_inode(inode_idx)`, which reads the inode directly from the image without traversing the directory structure, and subsequently without reading the directory entries. This caused the attribute `ext4.Inode.file_type` to default to `ext4.InodeType.UNKNOWN` (`!= ext4.InodeType.DIRECTORY`).
